### PR TITLE
JDK-8261520: JDK-8261302 breaks runtime/NMT/CheckForProperDetailStackTrace.java

### PR DESCRIPTION
**Pull request by tstuefe**
Since JDK-8261302, the test runtime/NMT/CheckForProperDetailStackTrace.java fails with
```
java.lang.RuntimeException: 'NativeCallStack::NativeCallStack' found in stdout
```

--

`NativeCallStack` contains a hash code. Before JDK-8261302, that hash code was calculated lazily in a non-inline hashcode getter. With JDK-8261302, the hash code calculation was moved into the `NativeCallStack` constructor and the getter was made inline.

The `NativeCallStack` constructor fills itself via `os::get_native_stack()`. Before JDK-8261302, that call has been the last call in the constructor and hence had been sometimes optimized into a tail call. Whether or not its a tail call matters since it affects the number of stack frames the stack walker has to skip. Therefore, the constructor contains coding to predict tail-call-ness:

```
#if (defined(_NMT_NOINLINE_) || defined(_WINDOWS) || !defined(_LP64) || defined(PPC64))
  // Not a tail call.
  toSkip++;
#if (defined(_NMT_NOINLINE_) && defined(BSD) && defined(_LP64))
  // Mac OS X slowdebug builds have this odd behavior where NativeCallStack::NativeCallStack
  // appears as two frames, so we need to skip an extra frame.
  toSkip++;
#endif // Special-case for BSD.
#endif // Not a tail call.
```

This prediction was now off since the hash code calculation happened at the end of the callstack. This causes the test error, since on some platforms (eg Linux x64) we now think we have a tail call when we don't, which means we do not skip enough frames, and the NMT output contains call frames like "NativeCallStack::NativeCallStack()", which trips the test.

-----------

Fix:

This fix moves the hash code calculation completely out of NativeCallStack. There is no reason why NativeCallStack should have a hash code. It mainly exists as a convenience to place it in a hash map. The patch moves the hash code calculation up into MallocSiteTableEntry.

This has the advantage of only having to pay for a hash code when you need it - in theory, one may use NativeCallStack in places other than NMT, where it is unnecessary. 

I considered other options:
- modify `os::get_native_stack()` to also calculate a hash in addition to capturing the stack, and return it in a caller provided variable. That would have left this call to be the tail call. However, it seemed less clean - we have two implementations of this function, as well as other, non-capturing, NativeCallStack constructors, which would have to be modified. It also would have made `os::get_native_stack()` less general purpose.
- Leave it as it is and just always skip frames: Seemed attractive, but I did not want to touch the tailcode-prediction-code and play whack-the-mole with platform specific test errors.

---------------

Tests: GA, manual test, nightlies at SAP

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261520](https://bugs.openjdk.java.net/browse/JDK-8261520): JDK-8261302 breaks runtime/NMT/CheckForProperDetailStackTrace.java

### Reviewers
 * [Zhengyu Gu](https://openjdk.java.net/census#zgu) (@zhengyu123 - **Reviewer**) ⚠️ Review applies to 78115d726a17fb4f0281e5bbbd0a73c0ec5d2bbc
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2672/head:pull/2672`
`$ git checkout pull/2672`

### CODE REVIEW

File: src/hotspot/share/services/mallocSiteTable.hpp [hunk 1]
```diff
@@ -57,12 +57,13 @@ class MallocSite : public AllocationSite {
 class MallocSiteHashtableEntry : public CHeapObj<mtNMT> {
  private:
   MallocSite                         _malloc_site;
+  const unsigned                     _hash;
```
Comment (zhengyu123) [inline src/hotspot/share/services/mallocSiteTable.hpp:60]: Prefer "unsigned int" to be consist with other places. Otherwise, Looks good to me.

**Review by zhengyu123 — APPROVED** [review]

File: src/hotspot/share/services/mallocSiteTable.hpp (hunk 1 above)
Comment (tstuefe) [reply src/hotspot/share/services/mallocSiteTable.hpp]: Thank you Zhengyu!

I reverted to unsigned int. Using just unsigned is a bad habit of mine, sorry.

Cheers, Thomas

**Review by coleenp — APPROVED** [review]
Yes this makes sense.  I use NativeCallStack sometimes for debugging things and don't need the hash code, so this is good.

**Comment by tstuefe** [comment]
> Yes this makes sense. I use NativeCallStack sometimes for debugging things and don't need the hash code, so this is good.

Yes, I do too. Also, I had some vague thoughts once about using it in UL to provide callstacks at log sites (was actually Robins idea).

Thanks!

**Pull request closed without merging**